# HG changeset patch
# User Pantxo Diribarne <pantxo.diribarne@gmail.com>
# Date 1538817036 -7200
#      Sat Oct 06 11:10:36 2018 +0200
# Node ID bb9fcc4525804cca244b03ad628ace2048186b7a
# Parent  b18f45eec7c81933559f6d88377dc8421cb5f3bc
Fix OpenGL errors when using getframe with non "pixels" figure units.

* getframe.m: Make sure the figure "position" is expressed in pixels as it is
  assumed further down __get_frame__ function.

diff --git a/scripts/image/getframe.m b/scripts/image/getframe.m
--- a/scripts/image/getframe.m
+++ b/scripts/image/getframe.m
@@ -94,17 +94,26 @@ function frame = getframe (h = [], rect 
     endif
     pos = rect;
   endif
 
   if (strcmp (get (hf, "visible"), "on")
       || (strcmp (get (hf, "__graphics_toolkit__"), "qt")
           && (strcmp (get (hf, "__gl_window__"), "on")
               || __have_feature__ ("QT_OFFSCREEN"))))
-    cdata = __get_frame__ (hf);
+    
+    ## __get_frame__ requires that the figure "units" is "pixels"
+    unwind_protect
+      units = get (hf, "units");
+      set (hf, "units", "pixels");
+      cdata = __get_frame__ (hf);
+    unwind_protect_cleanup
+      set (hf, "units", units)
+    end_unwind_protect
+    
   else
     error ("getframe: figure must be visible or qt toolkit must be used with __gl_window__ property 'on' or QT_OFFSCREEN feature available");
   endif
 
   i1 = max (floor (pos(1)), 1);
   i2 = min (ceil (pos(1)+pos(3)-1), columns (cdata));
   idxx = i1:i2;
   i1 = max (floor (pos(2)), 1);
